Backward compat tests (#979)
quickwit-inc/quickwit: quickwit-metastore/src/backward_compatibility_tests/README.md

# Backward compatibility test project.

This is just a project used to test backward compatibility of Quickwit.

> Right now, SplitMetadata, IndexMetadata and FileBasedMetadata are tested.

Reviewer: ```suggestion
Right now, `SplitMetadata`, `IndexMetadata`, and FileBackedIndex` are tested.
```

We want to be able to read all past versions of these files, but only write the most recent format.

> The tests consists in pairs of json file.

Reviewer: ```suggestion
The tests consist of pairs of JSON files.
```

XXXX.json and XXXX.expected.json

> XXXX.json consists in a json file in an old format.

Reviewer: ```suggestion
XXXX.json consists of a JSON file in an old format.
```

> XXXX.expected.json consists in the expected result of

Reviewer: ```suggestion
XXXX.expected.json consists of the expected result of
```

serialize_new_version(deserialize(XXXX.json)).

## Change of format autodetection

> Two things need to happen upon a change of format

Reviewer: ```suggestion
Two things need to happen upon a change of format.
```

### Updating expected.json

Of course to make all of this work, we need to keep `*.expected.json` files up-to-date
with the changes of format.

This is done in a semi-automatic fashion.

On a first pass, `deserialize(original_json) == deserialize(expectation_json)`.

> On a second pass, the tests are checking that `expectation_json = serialize(derialize(expectation_json))`.

Reviewer: ```suggestion
On a second pass, the tests are checking that `expectation_json = serialize(deserialize(expectation_json))`.
```

When changing the json format, it is expected to see this test fail.
The unit test then updates automatically the expected.json. The developer just has to
check the diff of the result (in particular no information should be lost) and commit the updated expected.json
files.

Adding this update operation within the unit test is a tad unexpected, but it has the merit of
integrating well with CI. If a developer forgets to update the expected.json file,
the CI will catch it.

### Adding a new test case.

If the serialization format changes, the unit test will also automatically add
a new unit test generated from a sample tested object.
Concretely, it will just write two files XXXX.json and XXXX.expected.json.

The two files will be identical. This is expected as this is a unit test for the
most recent version.

The unit test will start making sense in future updates thanks to the update phase
described in the previous section.

## How do I update `{IndexMetadata, SplitMetadata, FileBackedIndex}`?

> There two ways to types of upgrade.

Reviewer: ```suggestion
There are two types of upgrades.
```
